rust/main/agents/relayer/src/relayer/origin.rs

```diff
@@ -0,0 +1,361 @@
+use crate::settings::GasPaymentEnforcementConf;
+
+type MessageSync = Arc<dyn ContractSyncer<HyperlaneMessage>>;
+type InterchainGasPaymentSync = Arc<dyn ContractSyncer<InterchainGasPayment>>;
+type MerkleTreeHookSync = Arc<dyn ContractSyncer<MerkleTreeInsertion>>;
+
+#[derive(Clone)]
+pub struct Origin {
+    pub database: HyperlaneRocksDB,
+    pub domain: HyperlaneDomain,
+    pub index_settings: IndexSettings,
+    pub validator_announce: Arc<dyn ValidatorAnnounce>,
+    pub gas_payment_enforcer: Arc<RwLock<GasPaymentEnforcer>>,
+    pub prover_sync: Arc<RwLock<MerkleTreeBuilder>>,
+    pub message_sync: MessageSync,
+    pub interchain_gas_payment_sync: Option<InterchainGasPaymentSync>,
+    pub merkle_tree_hook_sync: MerkleTreeHookSync,
+}
+
+impl std::fmt::Debug for Origin {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Origin {{ domain: {} }}", self.domain.name())
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum FactoryError {
+    #[error("Failed to create validator announce for domain {0}: {1}")]
+    ValidatorAnnounce(String, String),
+    #[error("Failed to create message sync for domain {0}: {1}")]
+    MessageSync(String, String),
+    #[error("Failed to create igp sync for domain {0}: {1}")]
+    InterchainGasPaymentSync(String, String),
+    #[error("Failed to create merkle tree hook sync for domain {0}: {1}")]
+    MerkleTreeHookSync(String, String),
+}
+
+pub trait Factory {
+    async fn create(
+        &self,
+        domain: HyperlaneDomain,
+        chain_conf: &ChainConf,
+        gas_payment_enforcement: Vec<GasPaymentEnforcementConf>,
+    ) -> Result<Origin, FactoryError>;
+}
+
+#[derive(Clone, Debug)]
+pub struct OriginFactory {
+    db: DB,
+    core_metrics: Arc<CoreMetrics>,
+    sync_metrics: Arc<ContractSyncMetrics>,
+    advanced_log_meta: bool,
+    tx_id_indexing_enabled: bool,
+    igp_indexing_enabled: bool,
+}
+
+impl OriginFactory {
+    pub fn new(
+        db: DB,
+        core_metrics: Arc<CoreMetrics>,
+        sync_metrics: Arc<ContractSyncMetrics>,
+        advanced_log_meta: bool,
+        tx_id_indexing_enabled: bool,
+        igp_indexing_enabled: bool,
+    ) -> Self {
+        Self {
+            db,
+            core_metrics,
+            sync_metrics,
+            advanced_log_meta,
+            tx_id_indexing_enabled,
+            igp_indexing_enabled,
+        }
+    }
+}
+
+impl Factory for OriginFactory {
+    async fn create(
+        &self,
+        domain: HyperlaneDomain,
+        chain_conf: &ChainConf,
+        gas_payment_enforcement: Vec<GasPaymentEnforcementConf>,
+    ) -> Result<Origin, FactoryError> {
+        let db = HyperlaneRocksDB::new(&domain, self.db.clone());
+
+        let start_entity_init = Instant::now();
+        let validator_announce = self.init_validator_announce(chain_conf, &domain).await?;
+        self.measure(&domain, "validator_announce", start_entity_init.elapsed());
```

Comment:
It will be cleaner if we move measuring metrics inside each function.

Reply:
But some functions don't take in `domain` as a parameter, so I would need to pass in `domain` just for measuring its init time